### Assignment 2 
### Moldir Kumarbek 
### BD-2005

In [1]:
import pandas as pd
import numpy as np
import re
from pprint import pprint
from scipy.spatial.distance import pdist, squareform

## Task 1.1. Open the file

In [2]:
fo = open('words.txt', 'r')
my_list = fo.readlines()
print(my_list)

['In comparison to dogs, cats have not undergone major changes during the domestication process.\n', 'As cat simply catenates streams of bytes, it can be also used to concatenate binary files, where it will just concatenate sequence of bytes.\n', 'A common interactive use of cat for a single file is to output the content of a file to standard output.\n', 'Cats can hear sounds too faint or too high in frequency for human ears, such as those made by mice and other small animals.\n', 'In one, people deliberately tamed cats in a process of artificial selection, as they were useful predators of vermin.\n', 'The domesticated cat and its closest wild ancestor are both diploid organisms that possess 38 chromosomes and roughly 20,000 genes.\n', 'Domestic cats are similar in size to the other members of the genus Felis, typically weighing between 4 and 5 kg (8.8 and 11.0 lb).\n', 'However, if the output is piped or redirected, cat is unnecessary.\n', 'cat with one named file is safer where human

## Task 1.2. Each line is the one sentence. Please make them all in lower case form using string function lower()

In [3]:
with open('words.txt', 'r') as inp:
   y = inp.read().lower()
with open('words.txt', 'a') as out:
    out.write(y)

In [4]:
print(y)

in comparison to dogs, cats have not undergone major changes during the domestication process.
as cat simply catenates streams of bytes, it can be also used to concatenate binary files, where it will just concatenate sequence of bytes.
a common interactive use of cat for a single file is to output the content of a file to standard output.
cats can hear sounds too faint or too high in frequency for human ears, such as those made by mice and other small animals.
in one, people deliberately tamed cats in a process of artificial selection, as they were useful predators of vermin.
the domesticated cat and its closest wild ancestor are both diploid organisms that possess 38 chromosomes and roughly 20,000 genes.
domestic cats are similar in size to the other members of the genus felis, typically weighing between 4 and 5 kg (8.8 and 11.0 lb).
however, if the output is piped or redirected, cat is unnecessary.
cat with one named file is safer where human error is a concern - one wrong use of the

## Task 1.3. Tokenization

In [5]:
list1 = re.split('[^a-z]',y)

In [6]:
not_empty = [string for string in list1 if string != ""]
print(not_empty)

['in', 'comparison', 'to', 'dogs', 'cats', 'have', 'not', 'undergone', 'major', 'changes', 'during', 'the', 'domestication', 'process', 'as', 'cat', 'simply', 'catenates', 'streams', 'of', 'bytes', 'it', 'can', 'be', 'also', 'used', 'to', 'concatenate', 'binary', 'files', 'where', 'it', 'will', 'just', 'concatenate', 'sequence', 'of', 'bytes', 'a', 'common', 'interactive', 'use', 'of', 'cat', 'for', 'a', 'single', 'file', 'is', 'to', 'output', 'the', 'content', 'of', 'a', 'file', 'to', 'standard', 'output', 'cats', 'can', 'hear', 'sounds', 'too', 'faint', 'or', 'too', 'high', 'in', 'frequency', 'for', 'human', 'ears', 'such', 'as', 'those', 'made', 'by', 'mice', 'and', 'other', 'small', 'animals', 'in', 'one', 'people', 'deliberately', 'tamed', 'cats', 'in', 'a', 'process', 'of', 'artificial', 'selection', 'as', 'they', 'were', 'useful', 'predators', 'of', 'vermin', 'the', 'domesticated', 'cat', 'and', 'its', 'closest', 'wild', 'ancestor', 'are', 'both', 'diploid', 'organisms', 'that',

## Task 1.4. Make a list of all the words that appear in the sentences. 
- Note: all the words are unique. 
- Hint: we have 254 unique words.

In [7]:
myset = set(not_empty)
print({k: i for k, i in enumerate(sorted(myset))})

{0: 'a', 1: 'according', 2: 'adjacent', 3: 'allow', 4: 'allows', 5: 'also', 6: 'an', 7: 'ancestor', 8: 'and', 9: 'animals', 10: 'any', 11: 'app', 12: 'apple', 13: 'are', 14: 'arguments', 15: 'artificial', 16: 'as', 17: 'available', 18: 'based', 19: 'basic', 20: 'be', 21: 'between', 22: 'binary', 23: 'both', 24: 'by', 25: 'bytes', 26: 'can', 27: 'cat', 28: 'catenates', 29: 'cats', 30: 'changes', 31: 'chromosomes', 32: 'clear', 33: 'closest', 34: 'command', 35: 'commands', 36: 'common', 37: 'community', 38: 'comparison', 39: 'computers', 40: 'concatenate', 41: 'concern', 42: 'connected', 43: 'contains', 44: 'content', 45: 'count', 46: 'create', 47: 'default', 48: 'delete', 49: 'deliberately', 50: 'developed', 51: 'diploid', 52: 'disk', 53: 'displays', 54: 'dogs', 55: 'domestic', 56: 'domesticated', 57: 'domestication', 58: 'download', 59: 'drive', 60: 'during', 61: 'ears', 62: 'editions', 63: 'enhancements', 64: 'entirely', 65: 'error', 66: 'every', 67: 'external', 68: 'factory', 69: 'fa

## Task 1.5. Create Matrix with N x D dimensions. 
- N is the number of the sentences and D is the number of the unique words (22 x 254). 
- Fill it in: the element with index (i, j) in this matrix must be equal to the number of occurrences of the j-th word in the i-th sentence.

In [8]:
a = [[sentence.count(word) for word in myset] for sentence in my_list]

In [9]:
matrix = np.matrix(a)
matrix

matrix([[0, 0, 1, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 1, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]])

In [10]:
matrix.shape

(22, 254)

In [11]:
out = pd.DataFrame(matrix, index = my_list, columns=myset)
out.head()

non  are  cats  \
In comparison to dogs, cats have not undergone ...    0    0     1   
As cat simply catenates streams of bytes, it ca...    0    0     0   
A common interactive use of cat for a single fi...    0    0     0   
Cats can hear sounds too faint or too high in f...    0    0     0   
In one, people deliberately tamed cats in a pro...    0    0     1   

                                                    unnecessary  contains  \
In comparison to dogs, cats have not undergone ...            0         0   
As cat simply catenates streams of bytes, it ca...            0         0   
A common interactive use of cat for a single fi...            0         0   
Cats can hear sounds too faint or too high in f...            0         0   
In one, people deliberately tamed cats in a pro...            0         0   

                                                    app  three  ancestor  \
In comparison to dogs, cats have not undergone ...    0      0         0   
As cat simply catenates streams of bytes, it ca...    0      0         0   
A common interactive use of cat for a single fi...    0      0         0   
Cats can hear sounds too faint or too high in f...    0      0         0   
In one, people deliberately tamed cats in a pro...    0      0         0   

                                                    drive  other  ...  os  \
In comparison to dogs, cats have not undergone ...      0      0  ...   0   
As cat simply catenates streams of bytes, it ca...      0      0  ...   0   
A common interactive use of cat for a single fi...      0      0  ...   0   
Cats can hear sounds too faint or too high in f...      0      1  ...   1   
In one, people deliberately tamed cats in a pro...      0      0  ...   0   

                                                    interactive  diploid  \
In comparison to dogs, cats have not undergone ...            0        0   
As cat simply catenates streams of bytes, it ca...            0        0   
A common interactive use of cat for a single fi...            1        0   
Cats can hear sounds too faint or too high in f...            0        0   
In one, people deliberately tamed cats in a pro...            0        0   

                                                    information  installed  \
In comparison to dogs, cats have not undergone ...            0          0   
As cat simply catenates streams of bytes, it ca...            0          0   
A common interactive use of cat for a single fi...            0          0   
Cats can hear sounds too faint or too high in f...            0          0   
In one, people deliberately tamed cats in a pro...            0          0   

                                                    allow  installs  human  \
In comparison to dogs, cats have not undergone ...      0         0      0   
As cat simply catenates streams of bytes, it ca...      0         0      0   
A common interactive use of cat for a single fi...      0         0      0   
Cats can hear sounds too faint or too high in f...      0         0      1   
In one, people deliberately tamed cats in a pro...      0         0      0   

                                                    simply  may  
In comparison to dogs, cats have not undergone ...       0    0  
As cat simply catenates streams of bytes, it ca...       1    0  
A common interactive use of cat for a single fi...       0    0  
Cats can hear sounds too faint or too high in f...       0    0  
In one, people deliberately tamed cats in a pro...       0    0  

[5 rows x 254 columns]

## Task 1.6. Find the cosine distance from first sentence to the all other sentences. 
- Which two sentences is close to the first sentence. 
- You can use scipy.spatial.distance.cosine

In [12]:
from scipy.spatial.distance import pdist, squareform
#numpy array of my data
print(out.values)

[[0 0 1 ... 0 0 0]
 [0 0 0 ... 0 1 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [13]:
cosine_distances = pd.DataFrame(squareform(pdist(out.values, metric='cosine')), columns=out.index.values, index=out.index.values)
cosine_distances.head(1)

In comparison to dogs, cats have not undergone major changes during the domestication process.\n  \
In comparison to dogs, cats have not undergone ...                                                0.0                                                  

                                                    As cat simply catenates streams of bytes, it can be also used to concatenate binary files, where it will just concatenate sequence of bytes.\n  \
In comparison to dogs, cats have not undergone ...                                           0.112368                                                                                                

                                                    A common interactive use of cat for a single file is to output the content of a file to standard output.\n  \
In comparison to dogs, cats have not undergone ...                                           0.209901                                                            

                                                    Cats can hear sounds too faint or too high in frequency for human ears, such as those made by mice and other small animals.\n  \
In comparison to dogs, cats have not undergone ...                                            0.17181                                                                               

                                                    In one, people deliberately tamed cats in a process of artificial selection, as they were useful predators of vermin.\n  \
In comparison to dogs, cats have not undergone ...                                           0.108444                                                                         

                                                    The domesticated cat and its closest wild ancestor are both diploid organisms that possess 38 chromosomes and roughly 20,000 genes.\n  \
In comparison to dogs, cats have not undergone ...                                           0.138399                                                                                       

                                                    Domestic cats are similar in size to the other members of the genus Felis, typically weighing between 4 and 5 kg (8.8 and 11.0 lb).\n  \
In comparison to dogs, cats have not undergone ...                                           0.155061                                                                                       

                                                    However, if the output is piped or redirected, cat is unnecessary.\n  \
In comparison to dogs, cats have not undergone ...                                           0.191444                      

                                                    cat with one named file is safer where human error is a concern - one wrong use of the default redirection symbol ">" instead of "<" (often adjacent on keyboards) may permanently delete the file you were just needing to read.\n  \
In comparison to dogs, cats have not undergone ...                                           0.123526                                                                                                                                                                                     

                                                    In terms of legibility, a sequence of commands starting with cat and connected by pipes has a clear left-to-right flow of information.\n  \
In comparison to dogs, cats have not undergone ...                                           0.166989                                                                                          

                                                    ...  \
In comparison to dogs, cats have not undergone ...  ...   

                                                    When you type simply cat command without any arguments, it just receives the stdin content and displays it in the stdout.\n  \
In comparison to dogs, cats have not undergone ...    

In [14]:
cos = cosine_distances.iloc[0,:]
cosine = pd.DataFrame(cos)
cosine

In comparison to dogs, cats have not undergone major changes during the domestication process.\n
In comparison to dogs, cats have not undergone ...                                           0.000000                                               
As cat simply catenates streams of bytes, it ca...                                           0.112368                                               
A common interactive use of cat for a single fi...                                           0.209901                                               
Cats can hear sounds too faint or too high in f...                                           0.171810                                               
In one, people deliberately tamed cats in a pro...                                           0.108444                                               
The domesticated cat and its closest wild ances...                                           0.138399                                               
Domestic cats are similar in size to the other ...                                           0.155061                                               
However, if the output is piped or redirected, ...                                           0.191444                                               
cat with one named file is safer where human er...                                           0.123526                                               
In terms of legibility, a sequence of commands ...                                           0.166989                                               
Cat command is one of the basic commands that y...                                           0.200000                                               
Using cat command, the lines received from stdi...                                           0.180347                                               
When you type simply cat command without any ar...                                           0.187923                                               
Leopard was released on October 26, 2007 as the...                                           0.138057                                               
According to Apple, Leopard contains over 300 c...                                           0.153244                                               
As of Mid 2010, some Apple computers have firmw...                                           0.159167                                               
Since Apple moved to using Intel processors in ...                                           0.118745                                               
OS X Mountain Lion was released on July 25, 201...                                           0.173204                                               
Apple has released a small patch for the three ...                                           0.131435                                               
The Mountain Lion release marks the second time...                                           0.182513                                               
Mac OS X Mountain Lion installs in place, so yo...                                           0.141416                                               
The fifth major update to Mac OS X, Leopard, co...                                           0.173246

- From the output above, we see that **5th** (with 0.108444 cosine distance) and **2nd**  (with 0.112368 cosine distance) **sentences is close to the 1st sentence.**

## Task 2. Do the same step from Task 1 (steps 1 - 4)
- In this task you will create Term Frequency — Inverse Document Frequency matrix. 
- Find the cosine distance from first sentence to the all other sentences. Which two sentences is close to the first sentence. You can use scipy.spatial.distance.cosine. Is the any difference from the result of the previous Task? Note: You should not to use any existing libraries for tdf/idf. All the steps similar to the previous example.

In [15]:
df = open('words.txt', 'r')
my_list2 = df.readlines()

In [16]:
with open('words.txt', 'r') as inp2:
    y2 = inp2.read().lower()
with open('words.txt', 'a') as out2:
    out2.write(y2)

In [26]:
sentence_list2 = []

for element in my_list:
    sentence_list2.append(element.strip())

print(sentence_list2)

['In comparison to dogs, cats have not undergone major changes during the domestication process.', 'As cat simply catenates streams of bytes, it can be also used to concatenate binary files, where it will just concatenate sequence of bytes.', 'A common interactive use of cat for a single file is to output the content of a file to standard output.', 'Cats can hear sounds too faint or too high in frequency for human ears, such as those made by mice and other small animals.', 'In one, people deliberately tamed cats in a process of artificial selection, as they were useful predators of vermin.', 'The domesticated cat and its closest wild ancestor are both diploid organisms that possess 38 chromosomes and roughly 20,000 genes.', 'Domestic cats are similar in size to the other members of the genus Felis, typically weighing between 4 and 5 kg (8.8 and 11.0 lb).', 'However, if the output is piped or redirected, cat is unnecessary.', 'cat with one named file is safer where human error is a conc

In [18]:
list2 = re.split('[^a-z]',y2)

In [27]:
for i in range(len(sentence_list2)):
    sentence_list2[i] = sentence_list2[i].lower()
print(sentence_list2)

['in comparison to dogs, cats have not undergone major changes during the domestication process.', 'as cat simply catenates streams of bytes, it can be also used to concatenate binary files, where it will just concatenate sequence of bytes.', 'a common interactive use of cat for a single file is to output the content of a file to standard output.', 'cats can hear sounds too faint or too high in frequency for human ears, such as those made by mice and other small animals.', 'in one, people deliberately tamed cats in a process of artificial selection, as they were useful predators of vermin.', 'the domesticated cat and its closest wild ancestor are both diploid organisms that possess 38 chromosomes and roughly 20,000 genes.', 'domestic cats are similar in size to the other members of the genus felis, typically weighing between 4 and 5 kg (8.8 and 11.0 lb).', 'however, if the output is piped or redirected, cat is unnecessary.', 'cat with one named file is safer where human error is a conc

In [28]:
words = [w.replace('-', '') for w in sentence_list2]
words2 = [w1.replace(',', '') for w1 in words]
words3 = [w2.replace('.', '') for w2 in words2]
words4 = [w3.replace('/', '') for w3 in words3]
words5 = [w4.replace('(', '') for w4 in words4]
words6 = [w5.replace('\'', '') for w5 in words5]
words7 = [w6.replace(')', '') for w6 in words6]
sentences = [w7.replace('\"', '') for w7 in words7]
sentences

['in comparison to dogs cats have not undergone major changes during the domestication process',
 'as cat simply catenates streams of bytes it can be also used to concatenate binary files where it will just concatenate sequence of bytes',
 'a common interactive use of cat for a single file is to output the content of a file to standard output',
 'cats can hear sounds too faint or too high in frequency for human ears such as those made by mice and other small animals',
 'in one people deliberately tamed cats in a process of artificial selection as they were useful predators of vermin',
 'the domesticated cat and its closest wild ancestor are both diploid organisms that possess 38 chromosomes and roughly 20000 genes',
 'domestic cats are similar in size to the other members of the genus felis typically weighing between 4 and 5 kg 88 and 110 lb',
 'however if the output is piped or redirected cat is unnecessary',
 'cat with one named file is safer where human error is a concern  one wrong

In [29]:
y = ' '.join([str(elem) for elem in sentence_list2])
list2 = re.split('[^a-z]', y)
print(list2)

['in', 'comparison', 'to', 'dogs', '', 'cats', 'have', 'not', 'undergone', 'major', 'changes', 'during', 'the', 'domestication', 'process', '', 'as', 'cat', 'simply', 'catenates', 'streams', 'of', 'bytes', '', 'it', 'can', 'be', 'also', 'used', 'to', 'concatenate', 'binary', 'files', '', 'where', 'it', 'will', 'just', 'concatenate', 'sequence', 'of', 'bytes', '', 'a', 'common', 'interactive', 'use', 'of', 'cat', 'for', 'a', 'single', 'file', 'is', 'to', 'output', 'the', 'content', 'of', 'a', 'file', 'to', 'standard', 'output', '', 'cats', 'can', 'hear', 'sounds', 'too', 'faint', 'or', 'too', 'high', 'in', 'frequency', 'for', 'human', 'ears', '', 'such', 'as', 'those', 'made', 'by', 'mice', 'and', 'other', 'small', 'animals', '', 'in', 'one', '', 'people', 'deliberately', 'tamed', 'cats', 'in', 'a', 'process', 'of', 'artificial', 'selection', '', 'as', 'they', 'were', 'useful', 'predators', 'of', 'vermin', '', 'the', 'domesticated', 'cat', 'and', 'its', 'closest', 'wild', 'ancestor', 'a

In [30]:
not_empty2 = [string for string in list2 if string != ""]
print(not_empty2)

['in', 'comparison', 'to', 'dogs', 'cats', 'have', 'not', 'undergone', 'major', 'changes', 'during', 'the', 'domestication', 'process', 'as', 'cat', 'simply', 'catenates', 'streams', 'of', 'bytes', 'it', 'can', 'be', 'also', 'used', 'to', 'concatenate', 'binary', 'files', 'where', 'it', 'will', 'just', 'concatenate', 'sequence', 'of', 'bytes', 'a', 'common', 'interactive', 'use', 'of', 'cat', 'for', 'a', 'single', 'file', 'is', 'to', 'output', 'the', 'content', 'of', 'a', 'file', 'to', 'standard', 'output', 'cats', 'can', 'hear', 'sounds', 'too', 'faint', 'or', 'too', 'high', 'in', 'frequency', 'for', 'human', 'ears', 'such', 'as', 'those', 'made', 'by', 'mice', 'and', 'other', 'small', 'animals', 'in', 'one', 'people', 'deliberately', 'tamed', 'cats', 'in', 'a', 'process', 'of', 'artificial', 'selection', 'as', 'they', 'were', 'useful', 'predators', 'of', 'vermin', 'the', 'domesticated', 'cat', 'and', 'its', 'closest', 'wild', 'ancestor', 'are', 'both', 'diploid', 'organisms', 'that',

In [31]:
myset2 = set(not_empty2)
list_comp = {k: i for k, i in enumerate(sorted(myset2))}
print(list_comp)

{0: 'a', 1: 'according', 2: 'adjacent', 3: 'allow', 4: 'allows', 5: 'also', 6: 'an', 7: 'ancestor', 8: 'and', 9: 'animals', 10: 'any', 11: 'app', 12: 'apple', 13: 'are', 14: 'arguments', 15: 'artificial', 16: 'as', 17: 'available', 18: 'based', 19: 'basic', 20: 'be', 21: 'between', 22: 'binary', 23: 'both', 24: 'by', 25: 'bytes', 26: 'can', 27: 'cat', 28: 'catenates', 29: 'cats', 30: 'changes', 31: 'chromosomes', 32: 'clear', 33: 'closest', 34: 'command', 35: 'commands', 36: 'common', 37: 'community', 38: 'comparison', 39: 'computers', 40: 'concatenate', 41: 'concern', 42: 'connected', 43: 'contains', 44: 'content', 45: 'count', 46: 'create', 47: 'default', 48: 'delete', 49: 'deliberately', 50: 'developed', 51: 'diploid', 52: 'disk', 53: 'displays', 54: 'dogs', 55: 'domestic', 56: 'domesticated', 57: 'domestication', 58: 'download', 59: 'drive', 60: 'during', 61: 'ears', 62: 'editions', 63: 'enhancements', 64: 'entirely', 65: 'error', 66: 'every', 67: 'external', 68: 'factory', 69: 'fa

In [32]:
#Finding Term Frequency for unique words
word_frequency = {i: not_empty2.count(i) for i in not_empty2}
print(word_frequency)

{'in': 11, 'comparison': 1, 'to': 14, 'dogs': 1, 'cats': 4, 'have': 2, 'not': 1, 'undergone': 1, 'major': 2, 'changes': 2, 'during': 1, 'the': 20, 'domestication': 1, 'process': 2, 'as': 6, 'cat': 10, 'simply': 2, 'catenates': 1, 'streams': 1, 'of': 19, 'bytes': 2, 'it': 4, 'can': 3, 'be': 3, 'also': 2, 'used': 1, 'concatenate': 2, 'binary': 1, 'files': 1, 'where': 2, 'will': 2, 'just': 3, 'sequence': 2, 'a': 13, 'common': 1, 'interactive': 1, 'use': 2, 'for': 4, 'single': 1, 'file': 5, 'is': 7, 'output': 3, 'content': 2, 'standard': 1, 'hear': 1, 'sounds': 1, 'too': 2, 'faint': 1, 'or': 3, 'high': 1, 'frequency': 1, 'human': 2, 'ears': 1, 'such': 2, 'those': 1, 'made': 1, 'by': 3, 'mice': 1, 'and': 14, 'other': 2, 'small': 2, 'animals': 1, 'one': 4, 'people': 1, 'deliberately': 1, 'tamed': 1, 'artificial': 1, 'selection': 1, 'they': 1, 'were': 2, 'useful': 1, 'predators': 1, 'vermin': 1, 'domesticated': 1, 'its': 2, 'closest': 1, 'wild': 1, 'ancestor': 1, 'are': 2, 'both': 1, 'diploid

In [33]:
def compute_idf(word):
#на вход берется слово, для которого считаем IDF
#и корпус документов в виде списка списков слов
        #количество документов, где встречается искомый термин
        #считается как генератор списков
        return np.log10(len(not_empty2)/sum([1.0 for i in not_empty2 if word in i]))

In [34]:
print(compute_idf('mac'))

1.906694111260769


## Task 3. Please run the Hierarchy Clustering algorithm for the Task 1 and Task 2. And plot the dendrogram
- Please explain your results. 
- NOTE: by default scipy.cluster.hierarchy it uses euclidean distance. You should change it to the cosine distance.

## Task 3.1. The Hierarchy Clustering algorithm for the Task 1

In [37]:
#create dataset
d = {"Points":["P1", "P2", "P3", "P4", "P5", "P6","P7", "P8", "P9", "P10", "P11", 
               "P12","P13", "P14", "P15", "P16", "P17", "P18","P19", "P20", "P21", "P22"],
     'X': out.index.values, 
     'Y': out.columns}
data = pd.DataFrame(data=d).set_index("Points")
print(data)


#vizualizing
fig, ax = plt.subplots()
data.plot('X', 'Y', kind='scatter', ax=ax)

for k, v in df.iterrows():
    ax.annotate(k, v,  fontsize=12)

ValueError: arrays must all be same length

In [25]:
import scipy.cluster.hierarchy as model
dend_max = model.dendrogram(model.linkage(out[['X', 'Y']], method='complete', metric = 'cosine'), labels=out.index)

/Users/aliluabk/opt/anaconda3/lib/python3.8/site-packages/numpy/core/_asarray.py:102: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  return array(a, dtype, copy=False, order=order)


TypeError: unhashable type: 'numpy.ndarray'

## Task 3.2. The Hierarchy Clustering algorithm for the Task 2